src/main/scala/monix/kafka/KafkaConsumer.scala

```diff
@@ -0,0 +1,140 @@
+import monix.execution.{Ack, Cancelable, Scheduler}
+import monix.reactive.{Observable, Observer}
+import monix.reactive.observers.Subscriber
+import org.apache.kafka.clients.consumer.{ConsumerRecord, ConsumerRecords, KafkaConsumer => ApacheKafkaConsumer}
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+import scala.collection.JavaConverters._
+
+/** An `Observable` implementation that reads messages from `Kafka`.
+  *
+  * You need to call `addTopicSubscription` in order for the Observable to start sending messages from that topic.
+  */
+final class KafkaConsumer[K, V] private
+  (config: KafkaConsumerConfig, ioScheduler: Scheduler)
+  (implicit K: Deserializer[K], V: Deserializer[V])
+  extends Observable[ConsumerRecord[K, V]] with StrictLogging {
+
+  def unsafeSubscribeFn(subscriber: Subscriber[ConsumerRecord[K, V]]): Cancelable = {
+    val source = underlying.onErrorHandleWith { ex =>
+      logger.error(
+        s"Unexpected error in KafkaConsumerObservable, retrying in ${config.retryBackoffTime}", ex)
+      // Applying configured delay before retry!
+      underlying.delaySubscription(config.retryBackoffTime)
+    }
+
+    source.unsafeSubscribeFn(subscriber)
+  }
+
+  def addTopicSubscription(topics: Set[String]): Unit = {
+    logger.info(s"Subscribing to topics : ${topics.mkString(",")}")
+    consumerRef.subscribe((currentSubscriptions() ++ topics).asJavaCollection)
+  }
+
+  /**
+    * Removes a subscription from a particular topic.
+    *
+    * @param topics the topics to be unsubscribed, if it is an `Set.empty` this method will unsubscribe from all topics.
+    */
+  def removeTopicSubscription(topics: Set[String]): Unit =
+    if(topics.nonEmpty) {
+      logger.info(s"Unsubscribing from topics : ${topics.mkString(",")}")
+      consumerRef.subscribe((currentSubscriptions() -- topics).asJavaCollection)
+    } else {
+      logger.info(s"Unsubscribing from topics : ${currentSubscriptions().mkString(",")}")
+      consumerRef.unsubscribe()
+    }
+
+  // Gets initialized on `subscribe`
+  private lazy val consumerRef = {
+    logger.info(s"Kafka consumer connecting to servers: ${config.servers.mkString(",")}")
+    new ApacheKafkaConsumer[K,V](config.toProperties, K.create(), V.create())
+  }
+
+  private def currentSubscriptions() =
+    consumerRef.subscription().asScala.toSet
+
+  private lazy val underlying =
+    Observable.unsafeCreate[ConsumerRecord[K, V]] { subscriber =>
+      val cancelable = SingleAssignmentCancelable()
+
+      // Forced asynchronous boundary, in order to not block
+      // the current thread!
+      ioScheduler.executeAsync {
+        cancelable := Observable
+          .fromStateAction[Set[String], ConsumerRecords[K, V]]{ subscriptions =>
+            val pollRecords =
+              if (subscriptions.nonEmpty)
+                consumerRef.poll(config.fetchMaxWaitTime.toMillis)
```

Comment:
This `poll()` call is blocking and needs to be executed on the `ioScheduler`. I know that you did an `executeAsync` above, but that's not enough, because `fromStateAction` is going to jump threads at some point and it is going to use the main Scheduler instead of ioScheduler. I propose a loop based on `Task`:

```scala
def loop(consumer: KafkaConsumer[K,V], out: Subscriber[ConsumerRecord[K,V]]): Task[Unit] = {
    val sendTask = Task.unsafeCreate[Ack] { (s, conn, cb) =>
      io.executeAsync {
        val ack = try {
          val next = consumer.poll(pollTimeoutMillis)
          Observer.feed(out, next.asScala)
        } catch {
          case NonFatal(ex) =>
            Future.failed(ex)
        }

        ack.syncOnComplete {
          case Success(value) =>
            cb.asyncOnSuccess(value)(s)
          case Failure(ex) =>
            cb.asyncOnError(ex)(s)
        }
      }
    }

    sendTask.flatMap {
      case Stop => Task.unit
      case Continue => loop(consumer, out)
    }
  }
```